### PR TITLE
Enable implicit-feedback recommendation via one-class matrix factorization

### PR DESCRIPTION
**Pull request by wschin**
This PR updates LIBMF used in ML.NET to the latest [official master](https://github.com/cjlin1/libmf) for adding a parallel coordinate descent method which solves one-class matrix factorization. It's a part of #1408 and the remaining tasks are OpenMP, SSE, and more formulations (current ML.NET treats LIBMF as a library for regression problem so that classification and ranking formulation are not allowed).

### CODE REVIEW

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 1]
```diff
@@ -91,6 +91,13 @@ public sealed class MatrixFactorizationTrainer : TrainerBase<MatrixFactorization
     {
         public sealed class Arguments
         {
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Loss function minimized for finding factor matrices. " +
+                "Two values are allowed, 0 or 12. The values \"0\" means traditional collaborative filtering problem with squared loss. " +
+                "The value \"12\" triggers one-class matrix factorization for implicit-feedback recommendation problem.")]
+            [TGUI(SuggestedSweeps = "0,12")]
+            [TlcModule.SweepableDiscreteParam("Fun", new object[] { 0, 12 })]
+            public int Fun = 0;
```
Comment (wschin) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Enum

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 2]
```diff
@@ -116,6 +123,18 @@ public sealed class Arguments
             [TlcModule.SweepableDiscreteParam("Eta", new object[] { 0.001f, 0.01f, 0.1f })]
             public double Eta = 0.1;
 
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Coefficient of negative entries' loss in one-class matrix factorization.")]
+            [TGUI(SuggestedSweeps = "1,0.01,0.0001,0.000001")]
+            [TlcModule.SweepableDiscreteParam("Eta", new object[] { 1f, 0.01f, 0.0001f, 0.000001f})]
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >Eta [](start = 47, length = 3)

copypaste is bad for you man

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 3]
```diff
@@ -116,6 +123,18 @@ public sealed class Arguments
...
+            public double Alpha = 0.1;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Desired negative entries' value in one-class matrix factorization. In one-class matrix factorization, " +
+                "all matrix values observed are one (which can be viewed as positive cases in binary classification) while unobserved values " +
+                "(which can be viewed as negative cases in binary classification) need to be specified manually using this option.")]
+            [TGUI(SuggestedSweeps = "0.000001,0,0001,0.01")]
+            [TlcModule.SweepableDiscreteParam("Eta", new object[] { 0.000001f, 0.0001f, 0.01f })]
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >Eta [](start = 47, length = 3)

copypaste is bad for you man

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 4]
```diff
@@ -193,10 +215,13 @@ public MatrixFactorizationTrainer(IHostEnvironment env, Arguments args) : base(e
             Host.CheckUserArg(args.Lambda > 0, nameof(args.Lambda), posError);
             Host.CheckUserArg(args.Eta > 0, nameof(args.Eta), posError);
 
+            _fun = args.Fun;
             _lambda = args.Lambda;
             _k = args.K;
             _iter = args.NumIterations;
             _eta = args.Eta;
+            _alpha = args.Eta;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >_alpha = args.Eta; [](start = 12, length = 18)

are you sure?

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs (hunk 1 above)
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Do you want to add test which will cover this?

Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Added. Thanks.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs (hunk 2 above)
Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Oops.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs (hunk 3 above)
Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Yes sir.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs (hunk 4 above)
Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Fixed.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 5]
```diff
@@ -89,8 +89,17 @@ namespace Microsoft.ML.Trainers
     public sealed class MatrixFactorizationTrainer : TrainerBase<MatrixFactorizationPredictor>,
         IEstimator<MatrixFactorizationPredictionTransformer>
     {
+        public enum LibMFLossFunctionType { SquareLossRegression=0, SquareLossOneClass=12 };
+
         public sealed class Arguments
         {
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Loss function minimized for finding factor matrices. " +
+                "Two values are allowed, 0 or 12. The values \"0\" means traditional collaborative filtering problem with squared loss. " +
+                "The value \"12\" triggers one-class matrix factorization for implicit-feedback recommendation problem.")]
```
Comment (sfilipi) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: please put this text in the XML comments of this argument. I know we have zip xml comments in args classes, but at least let's start doing so on the ones we add.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 6]
```diff
@@ -89,8 +89,17 @@ namespace Microsoft.ML.Trainers
     public sealed class MatrixFactorizationTrainer : TrainerBase<MatrixFactorizationPredictor>,
         IEstimator<MatrixFactorizationPredictionTransformer>
     {
+        public enum LibMFLossFunctionType { SquareLossRegression=0, SquareLossOneClass=12 };
```
Comment (sfilipi) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >=0 [](start = 64, length = 2)

spaces.

File: test/Microsoft.ML.Tests/TrainerEstimators/MatrixFactorizationTests.cs [hunk 7]
```diff
@@ -331,5 +331,99 @@ public void MatrixFactorizationInMemoryDataZeroBaseIndex()
...
+            // Convert the in-memory matrix into an IDataView so that ML.NET components can consume it.
+            var testDataView = ComponentCreation.CreateDataView(Env, testDataMatrix);
+
+            // Apply the trained model to the test data.
+            var testPrediction = model.Transform(testDataView);
+
+            double tolerance = Math.Pow(10, -6);
+            var testResults = new List<OneClassMatrixElementZeroBasedForScore>(testPrediction.AsEnumerable<OneClassMatrixElementZeroBasedForScore>(mlContext, false));
+            // Positive example (i.e., examples can be found in dataMatrix) is close to 1.
+            Assert.InRange(testResults[0].Score, 0.982391 - tolerance, 0.982391 + tolerance);
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/TrainerEstimators/MatrixFactorizationTests.cs]: >Assert.InRange [](start = 12, length = 14)

you care use CompareNumbersWithTolerance

Comment (wschin) [reply test/Microsoft.ML.Tests/TrainerEstimators/MatrixFactorizationTests.cs]: Sure.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs (hunk 6 above)
Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Yes!

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs (hunk 5 above)
Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: I am not following. Could you elaborate?

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 8]
```diff
@@ -89,8 +89,19 @@ namespace Microsoft.ML.Trainers
...
         public sealed class Arguments
         {
+            /// <summary>
+            /// Loss function minimized for finding factor matrices.  Two values are allowed, 0 or 12. The values 0 means traditional collaborative filtering
+            /// problem with squared loss. The value 12 triggers one-class matrix factorization for implicit-feedback recommendation problem.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Loss function minimized for finding factor matrices.")]
+            [TGUI(SuggestedSweeps = "0,12")]
+            [TlcModule.SweepableDiscreteParam("Fun", new object[] { LibMFLossFunctionType.SquareLossRegression, LibMFLossFunctionType.SquareLossOneClass })]
+            public LibMFLossFunctionType Fun = LibMFLossFunctionType.SquareLossRegression;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >Fun [](start = 41, length = 3)

Can we call it LossFunction? "Fun" name isn't fun.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 9]
```diff
@@ -89,8 +89,19 @@ namespace Microsoft.ML.Trainers
     public sealed class MatrixFactorizationTrainer : TrainerBase<MatrixFactorizationPredictor>,
         IEstimator<MatrixFactorizationPredictionTransformer>
     {
+        public enum LibMFLossFunctionType { SquareLossRegression = 0, SquareLossOneClass = 12 };
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >LibMFLossFunctionType [](start = 20, length = 21)

this is nested enum, you don't have to be so specific, can you just call it LossFunction or LossFunctionType?
LibMF prefix is a bit scary.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 10]
```diff
@@ -116,6 +127,24 @@ public sealed class Arguments
             [TlcModule.SweepableDiscreteParam("Eta", new object[] { 0.001f, 0.01f, 0.1f })]
             public double Eta = 0.1;
 
+            /// <summary>
+            /// Importance of unobserved (i.e., negative) entries' loss in one-class matrix factorization.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Importance of unobserved entries' loss in one-class matrix factorization.")]
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >Importance of unobserved entries [](start = 59, length = 32)

1 is super important and 0 is not important at all? Or it's opposite? Can you specify range requirement?
Also I don't see any checks in code for this parameter.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 11]
```diff
@@ -193,10 +225,13 @@ public MatrixFactorizationTrainer(IHostEnvironment env, Arguments args) : base(e
             Host.CheckUserArg(args.Lambda > 0, nameof(args.Lambda), posError);
             Host.CheckUserArg(args.Eta > 0, nameof(args.Eta), posError);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >CheckUserArg [](start = 17, length = 12)

Do you wanna add Checks for your new added parameters here?

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: You added check for Alpha, what about C? 
Is it also only positive value?

Can C and Alpha be greater than 1?

---
In reply to: [235105129](https://github.com/dotnet/machinelearning/pull/1664#discussion_r235105129) [](ancestors = 235105129)

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 12]
```diff
@@ -116,6 +127,28 @@ public sealed class Arguments
...
+            /// usually a small value so that the solver is able to find a factorization equally good to unobserved and observed
+            /// entries. If only 10000 observed entries present in a 200000-by-300000 training matrix, one can try Alpha = 10000 / (200000*300000).
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Importance of unobserved entries' loss in one-class matrix factorization.")]
+            [TGUI(SuggestedSweeps = "1,0.01,0.0001,0.000001")]
+            [TlcModule.SweepableDiscreteParam("Alpha", new object[] { 1f, 0.01f, 0.0001f, 0.000001f})]
+            public double Alpha = 0.1;
+
+            /// <summary>
+            /// Desired negative entries' value in one-class matrix factorization. In one-class matrix factorization, all matrix values observed are one
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: >' [](start = 40, length = 1)

what this apostrophe  doing here?

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 13]
```diff
@@ -116,6 +127,28 @@ public sealed class Arguments
...
+            /// (which can be viewed as positive cases in binary classification) while unobserved values (which can be viewed as negative cases in binary
+            /// classification) need to be specified manually using this option.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Desired negative entries' value in one-class matrix factorization")]
+            [TGUI(SuggestedSweeps = "0.000001,0,0001,0.01")]
+            [TlcModule.SweepableDiscreteParam("C", new object[] { 0.000001f, 0.0001f, 0.01f })]
+            public double C = 0.000001f;
```
[the remark this answers is not in the document]
Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Changed to 0.0001.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs (hunk 12 above)
Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Removed.

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs (hunk 10 above)
Comment (wschin) [reply src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Doc extended again. :)

**Pull request merged**